calyx/src/passes/register_unsharing.rs

```diff
@@ -0,0 +1,287 @@
+            cell
+        }
+    }
+
+    fn get_group(
+        &mut self,
+        builder: &Builder,
+        group_name: &ir::Id,
+    ) -> RRC<Group> {
+        if self.group_map.contains_key(group_name) {
+            self.group_map.get(group_name).unwrap().clone()
+        } else {
+            let group = builder.component.find_group(group_name).unwrap();
+            self.group_map.insert(group_name.clone(), group.clone());
+            group
+        }
+    }
+
+    fn rename(
+        &mut self,
+        builder: &mut Builder,
+        rename_list: &[(ir::Id, ir::Id, Vec<GroupOrInvoke>)],
+    ) {
+        let mut grp_map: RewriteMap<&ir::Id> = HashMap::new();
+        let mut invoke_map: RewriteMap<ir::Id> = HashMap::new();
+        for (new_name, old_name, grouplist) in rename_list {
+            for group_or_invoke in grouplist {
+                match group_or_invoke {
+                    GroupOrInvoke::Group(group) => {
+                        grp_map.entry(group).or_default().push((
+                            self.get_cell(builder, old_name),
+                            self.get_cell(builder, new_name),
+                        ))
+                    }
+                    GroupOrInvoke::Invoke(invoke) => {
+                        invoke_map.entry(invoke.clone()).or_default().push((
+                            self.get_cell(builder, old_name),
+                            self.get_cell(builder, new_name),
+                        ))
+                    }
+                }
+            }
+        }
+
+        for (grp, rename_cells) in grp_map {
+            let group = self.get_group(builder, grp);
+            let mut group_ref = group.borrow_mut();
+            builder.rename_port_uses(&rename_cells, &mut group_ref.assignments)
+        }
+
+        self.invoke_map = invoke_map;
+    }
+}
+
+impl Visitor for RegisterUnsharing {
+    fn start(
+        &mut self,
+        comp: &mut ir::Component,
+        _c: &LibrarySignatures,
+    ) -> VisResult {
+        self.bookkeeper.replace(Bookkeeper::new(comp));
+        let mut builder = Builder::from(comp, _c, false);
+
+        let rename_list = self
+            .bookkeeper
+            .as_mut()
+            .unwrap()
+            .create_new_regs(&mut builder);
+
+        self.bookkeeper
+            .as_mut()
+            .unwrap()
+            .rename(&mut builder, &rename_list);
+
+        // for (group, z) in &self.bookkeeper.as_ref().unwrap().analysis.reach {
+        //     println!("Group {:?}", group);
+        //     println!("  {:?}", z);
+        // }
+
+        // for (x, y) in &self
+        //     .bookkeeper
+        //     .as_ref()
+        //     .unwrap()
+        //     .analysis
+        //     .calculate_overlap(&builder.component.continuous_assignments)
+        // {
+        //     println!("Overlapping defns for {}", x);
+        //     for def in y {
+        //         println!("   {:?}\n", def);
+        //     }
+        // }
```

Comment:
Remove debug comments unless you think we should keep them to simplify debugging in the future.

Reply:
Oops, no I think I'll remove these. I can always refer back in time if I need them again.